### DISCUSSION TITLE
[RFC] useForm 支持默认不初始化

### DISCUSSION THREAD

**Discussion by crazyair** [ant-design/ant-design#48308] [RFCs]
# RFC

## Summary

useForm 支持默认不初始化

## Motivation

useForm 需要在 render 前初始化，但是 Form 可能需要某个判断，默认不加载，而加载后，也可能需要某个判断再卸载，但是 useForm 就只会加载一次，导致无法跟 Form 的加载卸载同步。

一般这种情况会推荐使用 ref 方式，但是这个用法总是没有 useForm 方便

## Detailed design

```tsx
const [form] = Form.useForm(undefined, { manual: true }) 
// 是否加载
form.isLoad 
// 加载(初始化 form 实例)
form.load()
// 卸载
form.unLoad()
```

## Drawbacks

无

## Alternatives

```tsx
const [form] = Form.useForm(undefined, { manual: true }) 
// Form 加载， form 实例才初始化，Form 卸载， form 实例卸载
{ isLoad && <Form form={form}  /> } 
```

## Unresolved questions

无

**Comment by crazyair** [comment]
适用于 Form + Modal 基础使用，不用再 `formRef.current?.resetFields()`，可以直接 `<Modal onCancel = {() => form.unLoad() } />`

**Reply by Wxh16144** [reply]
你这么说我发现我前段时间也遇到过这个问题，解决方案也是这个

<img width="1405" alt="image" src="https://github.com/ant-design/ant-design/assets/32004925/5e01f8d2-a2cd-47fc-9e84-c680a3aae457">

**Reply by crazyair** [reply]
resetFields 不能彻底清理 form 的数据，这个方法不能用

**Comment by afc163** [comment]
一般不应该自行维护 React 之外的组件生命周期。

**Reply by crazyair** [reply]
是参考的这个 https://ahooks.js.org/zh-CN/hooks/use-request/index#%E6%89%8B%E5%8A%A8%E8%A7%A6%E5%8F%91 就是默认不初始化，而是手动初始化。

**Comment by afc163** [comment]
> 但是 useForm 就只会加载一次，导致无法跟 Form 的加载卸载同步。

用 preserve ？

**Reply by crazyair** [reply]
之前我记得有个，useForm  的 form 如果如果没给 Form 传，或者 Form 没有加载，控制台会有个错误提示

**Comment by zombieJ** [comment]
可以拓展一下 `useForm`，让它接受一个激活事件。如果绑定到 Form 上后，则触发事件。这样就可以自己决定要不要 `resetFields` 了。

**Reply by crazyair** [reply]
resetFields 方法不能清除 form 实例的数据，比如当 Form 卸载后，执行 `console.log(form.getFieldsValue(true));` 还能拿到 values 的值
